src/cata_tiles.cpp

```diff
@@ -2179,8 +2182,14 @@ bool cata_tiles::draw_terrain( const tripoint &p, const lit_level ll, int &heigh
             // do something to get other terrain orientation values
         }
         const std::string &tname = t.id().str();
+        const tripoint abs_pos = g->m.getabs( p );
         if( g->m.check_seen_cache( p ) ) {
-            g->u.memorize_tile( g->m.getabs( p ), tname, subtile, rotation );
+            g->u.memorize_tile( abs_pos, tname, subtile, rotation );
+        } else {
+            const memorized_terrain_tile &cached_tile = g->u.get_memorized_tile( abs_pos );
+            if( cached_tile.subtile != subtile && cached_tile.tile == tname ) {
```

Comment:
`rotation` also affects appearance of tiles, so it should also be compared. Anyway, I think you can just memorize the tile without comparing it to the old value.

Also I'm not sure if it fully fixes #36609. Could you test this case:
1. Move to a place with low light-level and straight walls, so you can only see a 3x3 area around the character.
2. Move near the wall so you can see three wall tiles, which are now memorized.
3. Move away from the wall so you can't see it, move 3 steps along the wall's direction, and then move near the wall again, so you see another three wall tiles.
4. Does the six wall tiles look like this: `======` (what we want) or `==><==` (what we had)?

Reply:
Rotation does change the appearance of the tile, but I don't think rotation should be changed without resetting the cache at that index, whereas revealing more information about the map does change the subtile without telling the cache.

And I believe the tile needs to be compared to the old one because (as far as I can tell) only the top level "thing" is stored in map memory. Furniture, for example, could be overwritten by terrain, as only one can be remembered. So if something has been seen there, it can't be overwritten unless I know it's terrain.
An improvement I might actually be able to do is to just check if it's terrain and overwrite it.
Another option is to un-see the tiles being drawn before re-seeing them

And I had noticed that situation in the test would happen, but I would consider that a different bug, as that's unrelated to the map cache. You can also fix the memory by just sweeping over the disconnections, and it'll fix it. I can't think of a convenient way to fix that bug.